Test parameters containing semicolons are truncated

@CharliePoole in https://github.com/nunit/nunit/issues/1885#issuecomment-268872044:

> @jnm2 I'm assigning this issue to you. Is there already an nunit 3 adapter issue? If not there should be.

The work is simple, but getting the Cake build task to complete so I can test seems not to be.

When it tries to build `NUnit3TestAdapterInstall.csproj`:

```
C:\Users\Joseph\Source\Repos\nunit3-vs-adapter\packages\Microsoft.VSSDK.BuildTools.15.0.25907-RC2\tools\VSSDK\Microsoft.VsSDK.targets(602,5): error MSB4018: The "GetExtensionsPath" task failed unexpectedly.\r [C:\Users\Joseph\Source\Repos\nunit3-vs-adapter\src\NUnit3TestAdapterInstall\NUnit3TestAdapterInstall.csproj]
C:\Users\Joseph\Source\Repos\nunit3-vs-adapter\packages\Microsoft.VSSDK.BuildTools.15.0.25907-RC2\tools\VSSDK\Microsoft.VsSDK.targets(602,5): error MSB4018: System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.VisualStudio.Settings.15.0, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.\r [C:\Users\Joseph\Source\Repos\nunit3-vs-adapter\src\NUnit3TestAdapterInstall\NUnit3TestAdapterInstall.csproj]
C:\Users\Joseph\Source\Repos\nunit3-vs-adapter\packages\Microsoft.VSSDK.BuildTools.15.0.25907-RC2\tools\VSSDK\Microsoft.VsSDK.targets(602,5): error MSB4018: File name: 'Microsoft.VisualStudio.Settings.15.0, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'\r [C:\Users\Joseph\Source\Repos\nunit3-vs-adapter\src\NUnit3TestAdapterInstall\NUnit3TestAdapterInstall.csproj]
C:\Users\Joseph\Source\Repos\nunit3-vs-adapter\packages\Microsoft.VSSDK.BuildTools.15.0.25907-RC2\tools\VSSDK\Microsoft.VsSDK.targets(602,5): error MSB4018:    at Microsoft.VsSDK.Build.Tasks.GetExtensionsPath.Execute()\r [C:\Users\Joseph\Source\Repos\nunit3-vs-adapter\src\NUnit3TestAdapterInstall\NUnit3TestAdapterInstall.csproj]
C:\Users\Joseph\Source\Repos\nunit3-vs-adapter\packages\Microsoft.VSSDK.BuildTools.15.0.25907-RC2\tools\VSSDK\Microsoft.VsSDK.targets(602,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()\r [C:\Users\Joseph\Source\Repos\nunit3-vs-adapter\src\NUnit3TestAdapterInstall\NUnit3TestAdapterInstall.csproj]
C:\Users\Joseph\Source\Repos\nunit3-vs-adapter\packages\Microsoft.VSSDK.BuildTools.15.0.25907-RC2\tools\VSSDK\Microsoft.VsSDK.targets(602,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()\r [C:\Users\Joseph\Source\Repos\nunit3-vs-adapter\src\NUnit3TestAdapterInstall\NUnit3TestAdapterInstall.csproj]
C:\Users\Joseph\Source\Repos\nunit3-vs-adapter\packages\Microsoft.VSSDK.BuildTools.15.0.25907-RC2\tools\VSSDK\Microsoft.VsSDK.targets(602,5): error MSB4018: \r [C:\Users\Joseph\Source\Repos\nunit3-vs-adapter\src\NUnit3TestAdapterInstall\NUnit3TestAdapterInstall.csproj]
C:\Users\Joseph\Source\Repos\nunit3-vs-adapter\packages\Microsoft.VSSDK.BuildTools.15.0.25907-RC2\tools\VSSDK\Microsoft.VsSDK.targets(602,5): error MSB4018: WRN: Assembly binding logging is turned OFF.\r [C:\Users\Joseph\Source\Repos\nunit3-vs-adapter\src\NUnit3TestAdapterInstall\NUnit3TestAdapterInstall.csproj]
C:\Users\Joseph\Source\Repos\nunit3-vs-adapter\packages\Microsoft.VSSDK.BuildTools.15.0.25907-RC2\tools\VSSDK\Microsoft.VsSDK.targets(602,5): error MSB4018: To enable assembly bind failure logging, set the registry value [HKLM\Software\Microsoft\Fusion!EnableLog] (DWORD) to 1.\r [C:\Users\Joseph\Source\Repos\nunit3-vs-adapter\src\NUnit3TestAdapterInstall\NUnit3TestAdapterInstall.csproj]
C:\Users\Joseph\Source\Repos\nunit3-vs-adapter\packages\Microsoft.VSSDK.BuildTools.15.0.25907-RC2\tools\VSSDK\Microsoft.VsSDK.targets(602,5): error MSB4018: Note: There is some performance penalty associated with assembly bind failure logging.\r [C:\Users\Joseph\Source\Repos\nunit3-vs-adapter\src\NUnit3TestAdapterInstall\NUnit3TestAdapterInstall.csproj]
C:\Users\Joseph\Source\Repos\nunit3-vs-adapter\packages\Microsoft.VSSDK.BuildTools.15.0.25907-RC2\tools\VSSDK\Microsoft.VsSDK.targets(602,5): error MSB4018: To turn this feature off, remove the registry value [HKLM\Software\Microsoft\Fusion!EnableLog].\r [C:\Users\Joseph\Source\Repos\nunit3-vs-adapter\src\NUnit3TestAdapterInstall\NUnit3TestAdapterInstall.csproj]
C:\Users\Joseph\Source\Repos\nunit3-vs-adapter\packages\Microsoft.VSSDK.BuildTools.15.0.25907-RC2\tools\VSSDK\Microsoft.VsSDK.targets(602,5): error MSB4018:  [C:\Users\Joseph\Source\Repos\nunit3-vs-adapter\src\NUnit3TestAdapterInstall\NUnit3TestAdapterInstall.csproj]
```

I'm using latest VS2017 (15.0.26014.0) with workloads ".NET desktop development", "Desktop development with C++", "Web development", ".NET Core and Docker (Preview)", and "Visual Studio extension development."
I even did a full repair to make sure nothing was corrupted.

`NUnit3TestAdapterInstall.csproj` says minimum VS version is 15.0, so I know you must have been using VS2017.

@nunit/contributors Any wisdom on this?

--- a/src/NUnitTestAdapter/NUnitTestAdapter.cs
+++ b/src/NUnitTestAdapter/NUnitTestAdapter.cs
@@ -30,6 +30,7 @@ extern alias ENG;
 using TestEngineClass = ENG::NUnit.Engine.TestEngine;
 
 using System;
+using System.Collections.Generic;
 using System.Diagnostics;
 using System.IO;
 using System.Reflection;
@@ -184,19 +185,9 @@ namespace NUnit.VisualStudio.TestAdapter
                 package.Settings[PackageSettings.RandomSeed] = Settings.RandomSeed;
 
             if (Settings.TestProperties.Count > 0)
-            {
-                var sb = new StringBuilder();
-                var index = 0;
-                foreach(string name in Settings.TestProperties.Keys)
-                {
-                    if (index++ > 0) sb.Append(";");
-                    sb.AppendFormat("{0}={1}", name, Settings.TestProperties[name]);
-                }
+                SetTestParameters(package.Settings, Settings.TestProperties);
 
-                package.Settings[PackageSettings.TestParameters] = sb.ToString();
-            }
-
-            // Always run one assembly at a time in process in it's own domain
+            // Always run one assembly at a time in process in its own domain
             package.Settings[PackageSettings.ProcessModel] = "InProcess";
 
             if (Settings.DomainUsage != null)
@@ -222,6 +213,27 @@ namespace NUnit.VisualStudio.TestAdapter
             package.Settings[PackageSettings.WorkDirectory] = workDir;
 
             return package;
+        }
+
+        /// <summary>
+        /// Sets test parameters, handling backwards compatibility.
+        /// </summary>
+        private static void SetTestParameters(IDictionary<string, object> runSettings, IDictionary<string, string> testParameters)
+        {
+            runSettings[PackageSettings.TestParametersDictionary] = testParameters;
+
+            if (testParameters.Count != 0)
+            {
+                // Kept for backwards compatibility with old frameworks.
+                // Reserializes the way old frameworks understand, even if the parsing above is changed.
+                // This reserialization cannot be changed without breaking compatibility with old frameworks.
+
+                var oldFrameworkSerializedParameters = new StringBuilder();
+                foreach (var parameter in testParameters)
+                    oldFrameworkSerializedParameters.Append(parameter.Key).Append('=').Append(parameter.Value).Append(';');
+
+                runSettings[PackageSettings.TestParameters] = oldFrameworkSerializedParameters.ToString(0, oldFrameworkSerializedParameters.Length - 1);
+            }
         }
 
         protected static void CleanUpRegisteredChannels()

--- a/src/NUnitTestAdapter/PackageSettings.cs
+++ b/src/NUnitTestAdapter/PackageSettings.cs
@@ -199,9 +199,14 @@ namespace NUnit.Common
         public const string DefaultTestNamePattern = "DefaultTestNamePattern";
 
         /// <summary>
-        /// Parameters to be passed on to the test
+        /// Parameters to be passed on to the tests, serialized to a single string which needs parsing. Obsoleted by <see cref="TestParametersDictionary"/>; kept for backward compatibility.
         /// </summary>
         public const string TestParameters = "TestParameters";
+
+        /// <summary>
+        /// Parameters to be passed on to the tests, already parsed into an IDictionary&lt;string, string>. Replaces <see cref="TestParameters"/>.
+        /// </summary>
+        public const string TestParametersDictionary = "TestParametersDictionary";
 
         #endregion
 

